homeassistant/config_entries.py

```diff
@@ -0,0 +1,516 @@
+        entry = self._entries.pop(found)
+        self._async_schedule_save()
+
+        unloaded = yield from entry.async_unload(self.hass)
+
+        return {
+            'require_restart': not unloaded
+        }
+
+    @asyncio.coroutine
+    def async_load(self):
+        """Load the config."""
+        path = self.hass.config.path(PATH_CONFIG)
+        if not os.path.isfile(path):
+            self._entries = []
+            return
+
+        entries = yield from self.hass.async_add_job(load_json, path)
+        self._entries = [ConfigEntry(**entry) for entry in entries]
+
+    @asyncio.coroutine
+    def _async_add_entry(self, entry):
+        """Add an entry."""
+        self._entries.append(entry)
+        self._async_schedule_save()
+
+        # Setup entry
+        if entry.domain in self.hass.config.components:
+            # Component already set up, just need to call setup_entry
+            yield from entry.async_setup(self.hass)
+        else:
+            # Setting up component will also load the entries
+            yield from async_setup_component(
+                self.hass, entry.domain, self._hass_config)
+
+    @callback
+    def _async_schedule_save(self):
+        """Schedule saving the entity registry."""
+        if self._sched_save is not None:
+            self._sched_save.cancel()
+
+        self._sched_save = self.hass.loop.call_later(
+            SAVE_DELAY, self.hass.async_add_job, self._async_save
+        )
+
+    @asyncio.coroutine
+    def _async_save(self):
+        """Save the entity registry to a file."""
+        self._sched_save = None
+        data = [entry.as_dict() for entry in self._entries]
+
+        yield from self.hass.async_add_job(
+            save_json, self.hass.config.path(PATH_CONFIG), data)
+
+
+class FlowManager:
+    """Manage all the config flows that are in progress."""
+
+    def __init__(self, hass, hass_config, async_add_entry):
+        """Initialize the flow manager."""
+        self.hass = hass
+        self._hass_config = hass_config
+        self._progress = {}
+        self._async_add_entry = async_add_entry
+
+    @callback
+    def async_progress(self):
+        """Return the flows in progress."""
+        return [{
+            'flow_id': flow.flow_id,
+            'domain': flow.domain,
+            'source': flow.source,
+        } for flow in self._progress.values()]
+
+    @asyncio.coroutine
+    def async_init(self, domain, *, source=SOURCE_USER, data=None):
+        """Start a configuration flow."""
+        handler = HANDLERS.get(domain)
+
+        if handler is None:
+            # This will load the component and thus register the handler
+            component = getattr(self.hass.components, domain)
+            handler = HANDLERS.get(domain)
+
+            if handler is None:
+                raise self.hass.helpers.UnknownHandler
+
+            # Make sure requirements and dependencies of component are resolved
+            yield from async_process_deps_reqs(
```

Comment:
If I get this correctly, this means that HA will install all dependencies of a component as soon as the user clicks on "configure", right? Judging from myself, I'd assume that many curious users will click "configure" on many integrations, just to see what data is necessary there. If you happen to click on e.g. the Z-Wave integration, this will freeze your raspberry pi for an hour or so. I guess most users will assume something broke and restart HA.

How about this: Instead of installing dependencies here, add a possible `RESULT_TYPE_INSTALL_DEPENDENCIES`. That way, a component that needs its dependencies installed to provide the configuration flow (which is probably only a small subset of components? Why do you need your dependencies to show the config options?) can as first config step show a box saying "I need to install X dependencies. This will take a while. Do you want to continue?". The user clicks 'yes', the step returns `RESULT_TYPE_INSTALL_DEPENDENCIES`, we install the dependencies and continue with the next step.

Reply:
Oh, and obviously you'd need to install the dependencies after the config flow finished successfully if there was no `RESULT_TYPE_INSTALL_DEPENDENCIES`, because then the entity must be set up…

Reply:
The odds of a config flow needing access to the dependencies are high, as we will need to validate that the passed in info is correct. This requires making a connection with the info -> requires the dep.

I guess that we can show a spinner while the flow is "prepared" (loading dependencies and requirements).

Oh, you know, you're right. We will give flow developers access to trigger the loading/installation of dependencies/requirements. That way they can do it in step 2, 3 or whenever they need it. Do less and be lazy! (and thus do stuff as late as possible).

No need for an extra result type, the frontend should show a spinner while the next step is loading (this is a TODO btw). Installing dependencies will be done while loading. For things like Z-Wave, we can first show a page, "Welcome to setting up Z-Wave. Installation can take a while, please sit back."

Reply:
I will do this in a followup PR.